### DISCUSSION TITLE
api模式下是不能实现工具调用吗？

### DISCUSSION THREAD

**Discussion by 4ooooo** [zai-org/ChatGLM3#1250] [Q&A]
我在服务器上起server，然后在本地终端调用api，
response = client.chat.completions.create(
model="chatglm3_6b", # 填写需要调用的模型名称
messages=messages,
tools=tools,
tool_choice="auto",
)
这样是无法识别到工具的，请问有什么好的解决办法吗？需要通过api实现

**Comment by JefferyLJF1129** [comment]
Hi我也在尝试使用api实现工具调用，模型一直都不会去调用工具。不知道跟你是不是一样的情况~

**Reply by 4ooooo** [reply]
应该是的，github上给的api示例跑不通，这块卡了一天了，如果本地加载模型就能识别到工具清单。

**Reply by JefferyLJF1129** [reply]
我用阿里云部署了模型，起了api server之后用api示例跑通了。

**Reply by 4ooooo** [reply]
请问是哪个文件夹里的哪个文件  我试一下

**Reply by JefferyLJF1129** [reply]
先是openai_api_demo里api_server.py把服务run起来，然后运行tools_using_demo里的openai_api_demo.py

**Reply by 4ooooo** [reply]
谢谢！我一会儿试一下

**Reply by JefferyLJF1129** [reply]
不过我这个流程感觉本质还是在调本地模型T T 我之前一直都是服务器上部署然后终端调，都没通

**Reply by JefferyLJF1129** [reply]
你之前试的就是tools_using_demo里的代码吗？有啥修改不

**Reply by JefferyLJF1129** [reply]
方便问下你之前模型是部在哪儿么

**Reply by 4ooooo** [reply]
我是按照调用那个商用的glm4的代码写的，就是问题里的那段代码。

**Reply by 4ooooo** [reply]
autodl上租的服务器

**Reply by 4ooooo** [reply]
请问你运行openai_api_demo.py，模型选择了查询天气的那个工具了吗？我测试的回复是“ 我无法直接查询实时天气信息，但您可以通过访问中国气象局网站（http://www.cma.gov.cn/）或者使用手机上的天气应用来获取北京的实时天气信息。”

**Reply by JefferyLJF1129** [reply]
emmm这显然不是工具调用了……

**Reply by 4ooooo** [reply]
我知道，我想确认一下，看会不会是我pull的镜像的问题

**Reply by JefferyLJF1129** [reply]
拉的模型权重吗？还是啥 
我不知道会不会模型文件版本不一致，我在尝试用我在阿里云部署的那份模型文件用原本的调用方法复现出来，正在尝试中

你是使用啥镜像？这个会和镜像有关嘛？

**Reply by 4ooooo** [reply]
我用的是autodl算法社区里的一个，我想看是不是模型版本的问题，因为如果你用原版的能跑通，我用镜像不行，那可能就是模型的问题了

**Reply by JefferyLJF1129** [reply]
今天重新用了huggingface上的模型文件部署了一次，还是不能调工具

**Reply by 4ooooo** [reply]
我换成qwen7b可以了。但我刚看glm4发布了，新版可能可以了吧😂

**Reply by JefferyLJF1129** [reply]
方便问下跑的是哪个代码吗

**Reply by 4ooooo** [reply]
Qwen-7B-Chat，阿里通义千问的开源，直接用他github上的openai-api.py那个代码起服务，就可以了，能调用工具。GLM4我看还没有给出api_server的代码

**Reply by JefferyLJF1129** [reply]
大佬方便发一下qwen这个openai的链接吗

**Reply by JefferyLJF1129** [reply]
起服务的代码和调工具的代码

**Reply by 4ooooo** [reply]
https://github.com/QwenLM/Qwen/tree/main  main里的openai_api.py和example里的function_call.py
